File: src/main/java/rocks/huwi/walkingdinner/geneticplanner/CoursesProblem.java
package rocks.huwi.walkingdinner.geneticplanner;

import io.jenetics.EnumGene;
import io.jenetics.Genotype;
import io.jenetics.PermutationChromosome;
import io.jenetics.engine.Codec;
import io.jenetics.engine.Problem;
import io.jenetics.util.ISeq;
import rocks.huwi.walkingdinner.geneticplanner.dietcompatibility.CourseCompatibility;
import rocks.huwi.walkingdinner.geneticplanner.location.Location;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

class CoursesProblem implements Problem<Courses, EnumGene<Team>, Double> {
    private final ISeq<Team> teams;

    CoursesProblem(final ISeq<Team> teams) {
        this.teams = teams;
    }

    @Override
    public Codec<Courses, EnumGene<Team>> codec() {
        return Codec.of(
                Genotype.of(
                        PermutationChromosome.of(teams),
                        PermutationChromosome.of(teams),
                        PermutationChromosome.of(teams)),
                genotype -> new Courses(
                        genotype.getChromosome(0).toSeq().map(EnumGene::getAllele),
                        genotype.getChromosome(1).toSeq().map(EnumGene::getAllele),
                        genotype.getChromosome(2).toSeq().map(EnumGene::getAllele)
                )
        );
    }

    private void addLocations(HashMap<Team, List<Location>> teamsLocations, Set<Meeting> meetings) {
        for (Meeting meeting : meetings) {
            for (Team team : meeting.getTeams()) {
                List<Location> teamLocations = teamsLocations.computeIfAbsent(team, x -> new ArrayList<>());
                teamLocations.add(meeting.getCookingTeam().getLocation());
            }
        }
    }

    private double calculateLocationsDistance(List<Location> locations) {
        return locations.get(0).calculateDistance(locations.get(1)) +
                locations.get(1).calculateDistance(locations.get(2));
    }

    private double calculateOverallDistance(Courses courses) {
        Set<Meeting> meetings = courses.toMeetings();

        Map<String, Set<Meeting>> courseMeetings = meetings.stream()
                .collect(Collectors.groupingBy(Meeting::getCourse, Collectors.toSet()));

        Map<Team, List<Location>> teamsLocations = getTeamLocations(courseMeetings);

        return teamsLocations.entrySet().stream()
                .map(Map.Entry::getValue)
                .map(this::calculateLocationsDistance)
                .mapToDouble(d -> d)
                .sum();
    }

    private HashMap<Team, List<Location>> getTeamLocations(Map<String, Set<Meeting>> courseMeetings) {
        HashMap<Team, List<Location>> teamsLocations = new HashMap<>();

        addLocations(teamsLocations, courseMeetings.get(Courses.course1name));
        addLocations(teamsLocations, courseMeetings.get(Courses.course2name));
        addLocations(teamsLocations, courseMeetings.get(Courses.course3name));

        return teamsLocations;
    }

    private double calculateMultipleCookingTeams(Set<Meeting> meetings) {
        Map<Team, Long> teamCookings = meetings.stream()
                .map(Meeting::getCookingTeam)
                .collect(Collectors.groupingBy(t -> t, Collectors.counting()));
        Long countMultipleCookingTeams = teamCookings.entrySet().stream()
                .filter(kv -> kv.getValue() > 1)
                .map(Map.Entry::getValue)
                .mapToLong(value -> value)
                .sum();

        return countMultipleCookingTeams.doubleValue();
    }

    private double calculateIncompatibleTeams(Set<Meeting> meetings) {
        return ((Long) meetings.stream()
//                .filter(m -> !HardCompatibility.INSTANCE.areCompatibleTeams(m))
                .filter(m -> !CourseCompatibility.INSTANCE.areCompatibleTeams(m))
                .count())
                .doubleValue();
    }

    @Override
    public Function<Courses, Double> fitness() {
        return courses ->
        {
            Map<String, Set<Meeting>> courseMeetings = courses.toCourseMeetings();
            Set<Meeting> meetings = courseMeetings.entrySet().stream()
                    .flatMap(cm -> cm.getValue().stream())
                    .collect(Collectors.toSet());

            return (1 * calculateMultipleCookingTeams(meetings))
                    + (1 * calculateIncompatibleTeams(courseMeetings.get(Courses.course1name)))
                    + (1 * calculateIncompatibleTeams(courseMeetings.get(Courses.course2name)))
                    + (1 * calculateIncompatibleTeams(courseMeetings.get(Courses.course3name)))
                    + (0.00001 * calculateOverallDistance(courses));
        };
    }
}

File: planners/src/main/kotlin/de/debuglevel/walkingdinner/planner/geneticplanner/CoursesProblem.kt
package de.debuglevel.walkingdinner.planner.geneticplanner

import de.debuglevel.walkingdinner.model.Courses
import de.debuglevel.walkingdinner.model.Meeting
import de.debuglevel.walkingdinner.model.dietcompatibility.CourseCompatibility
import de.debuglevel.walkingdinner.model.location.Location
import de.debuglevel.walkingdinner.model.team.Team
import de.debuglevel.walkingdinner.planner.geneticplanner.CoursesProblemLegacyJavaCode.calculateMultipleCookingTeams
import de.debuglevel.walkingdinner.planner.geneticplanner.CoursesProblemLegacyJavaCode.calculateOverallDistance
import io.jenetics.EnumGene
import io.jenetics.Genotype
import io.jenetics.PermutationChromosome
import io.jenetics.engine.Codec
import io.jenetics.engine.Problem
import io.jenetics.util.ISeq
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors

class CoursesProblem(private val teams: ISeq<Team>) : Problem<Courses, EnumGene<Team>, Double> {
    override fun codec(): Codec<Courses, EnumGene<Team>> {

        val encoding: Genotype<EnumGene<Team>> = Genotype.of(
                PermutationChromosome.of(teams),
                PermutationChromosome.of(teams),
                PermutationChromosome.of(teams))

        val decoder: Function<Genotype<EnumGene<Team>>, Courses> = Function { genotype: Genotype<EnumGene<Team>> ->
            Courses(
                    genotype.getChromosome(0).toSeq().map { it.allele },
                    genotype.getChromosome(1).toSeq().map { it.allele },
                    genotype.getChromosome(2).toSeq().map { it.allele }
            )
        }

        val codec: Codec<Courses, EnumGene<Team>> = Codec.of<EnumGene<Team>, Courses>(
                encoding,
                decoder)

        return codec
    }

    companion object {
        fun calculateLocationsDistance(locations: List<Location>): Double {
            return locations[0].calculateDistance(locations[1]) + locations[1].calculateDistance(locations[2])
        }

        fun getTeamLocations(courseMeetings: Map<String, Set<Meeting>>): HashMap<Team, List<Location>> {
            val teamsLocations = HashMap<Team, List<Location>>()

            CoursesProblemLegacyJavaCode.addLocations(teamsLocations, courseMeetings[Courses.course1name])
            CoursesProblemLegacyJavaCode.addLocations(teamsLocations, courseMeetings[Courses.course2name])
            CoursesProblemLegacyJavaCode.addLocations(teamsLocations, courseMeetings[Courses.course3name])

            return teamsLocations
        }

        fun calculateIncompatibleTeams(meetings: Set<Meeting>): Double {
            return meetings.stream()
                    //                .filter(m -> !HardCompatibility.INSTANCE.areCompatibleTeams(m))
                    .filter { m -> !CourseCompatibility.areCompatibleTeams(m) }
                    .count()
                    .toDouble()
        }
    }

    override fun fitness(): Function<Courses, Double> {
        return Function { courses ->
            val courseMeetings = courses.toCourseMeetings()
            val meetings = courseMeetings.entries.stream()
                    .flatMap { cm -> cm.value.stream() }
                    .collect(Collectors.toSet<Meeting>())

            val d = (1 * calculateMultipleCookingTeams(meetings)
                    + 1 * calculateIncompatibleTeams(courseMeetings.getValue(Courses.course1name))
                    + 1 * calculateIncompatibleTeams(courseMeetings.getValue(Courses.course2name))
                    + 1 * calculateIncompatibleTeams(courseMeetings.getValue(Courses.course3name))
                    + 0.00001 * calculateOverallDistance(courses))

            d
        }
    }
}
